File: libraries/react-native-ondevice-ai/android/src/main/java/com/margelo/nitro/ondeviceai/OndeviceAiHelper.kt
package com.margelo.nitro.ondeviceai

import com.locanara.ChatMessageInput
import com.locanara.ChatParametersInput
import com.locanara.DeviceCapability
import com.locanara.FeatureType
import com.locanara.ProofreadInputType
import com.locanara.RewriteOutputType
import com.locanara.SummarizeInputType
import com.locanara.SummarizeOutputType

/** Decodes Nitro option structs into Locanara chain constructor parameters */
object OndeviceAiHelper {
    private val wrapperFeatures =
        setOf(
            FeatureType.SUMMARIZE,
            FeatureType.CLASSIFY,
            FeatureType.EXTRACT,
            FeatureType.CHAT,
            FeatureType.TRANSLATE,
            FeatureType.REWRITE,
            FeatureType.PROOFREAD,
        )
    private val promptCapabilities = setOf("chat", "classify", "extract", "translate")

    fun isWrapperReady(capability: DeviceCapability): Boolean = capability.availableFeatures.any { it in wrapperFeatures }

    fun isWrapperSupported(capability: DeviceCapability): Boolean {
        val promptSupported =
            capability.modelInfo
                ?.geminiNanoAndroid
                ?.capabilities
                ?.any { it in promptCapabilities } == true
        return isWrapperReady(capability) || promptSupported
    }

    // region Summarize

    fun summarizeOutputType(options: Variant_NullType_NitroSummarizeOptions?): SummarizeOutputType {
        val opts = options?.asSecondOrNull()
        return when (opts?.outputType) {
            NitroSummarizeOutputType.TWO_BULLETS -> SummarizeOutputType.TWO_BULLETS
            NitroSummarizeOutputType.THREE_BULLETS -> SummarizeOutputType.THREE_BULLETS
            else -> SummarizeOutputType.ONE_BULLET
        }
    }

    fun summarizeInputType(options: Variant_NullType_NitroSummarizeOptions?): SummarizeInputType {
        val opts = options?.asSecondOrNull()
        return when (opts?.inputType) {
            NitroSummarizeInputType.CONVERSATION -> SummarizeInputType.CONVERSATION
            else -> SummarizeInputType.ARTICLE
        }
    }

    // endregion

    // region Classify

    fun classifyOptions(options: Variant_NullType_NitroClassifyOptions?): Pair<List<String>, Int> {
        val opts = options?.asSecondOrNull()
        val categories =
            opts?.categories?.asSecondOrNull()?.toList()
                ?: listOf("positive", "negative", "neutral")
        val maxResults = opts?.maxResults?.asSecondOrNull()?.toInt() ?: 3
        return Pair(categories, maxResults)
    }

    // endregion

    // region Extract

    fun entityTypes(options: Variant_NullType_NitroExtractOptions?): List<String> {
        val opts = options?.asSecondOrNull()
        return opts?.entityTypes?.asSecondOrNull()?.toList()
            ?: listOf("person", "location", "date", "organization")
    }

    // endregion

    // region Chat

    fun chatParameters(options: Variant_NullType_NitroChatOptions?): ChatParametersInput {
        val opts = options?.asSecondOrNull()
        val systemPrompt =
            opts?.systemPrompt?.asSecondOrNull()
                ?: "You are a friendly, helpful assistant."

        val history =
            opts?.history?.asSecondOrNull()?.map { message ->
                ChatMessageInput(
                    role = message.role.name.lowercase(),
                    content = message.content,
                )
            }

        return ChatParametersInput(
            conversationId = opts?.conversationId?.asSecondOrNull(),
            systemPrompt = systemPrompt,
            history = history,
        )
    }

    // endregion

    // region Translate

    fun translateOptions(options: NitroTranslateOptions): Pair<String, String> = Pair(options.sourceLanguage, options.targetLanguage)

    // endregion

    // region Rewrite

    fun rewriteStyle(options: NitroRewriteOptions): RewriteOutputType =
        when (options.outputType) {
            NitroRewriteOutputType.ELABORATE -> RewriteOutputType.ELABORATE
            NitroRewriteOutputType.EMOJIFY -> RewriteOutputType.EMOJIFY
            NitroRewriteOutputType.SHORTEN -> RewriteOutputType.SHORTEN
            NitroRewriteOutputType.FRIENDLY -> RewriteOutputType.FRIENDLY
            NitroRewriteOutputType.PROFESSIONAL -> RewriteOutputType.PROFESSIONAL
            NitroRewriteOutputType.REPHRASE -> RewriteOutputType.REPHRASE
        }

    // endregion

    // region Proofread

    fun proofreadInputType(options: NitroProofreadOptions): ProofreadInputType =
        when (options.inputType) {
            NitroProofreadInputType.VOICE -> ProofreadInputType.VOICE
            NitroProofreadInputType.KEYBOARD -> ProofreadInputType.KEYBOARD
        }

    // endregion
}

File: libraries/react-native-ondevice-ai/ios/OndeviceAiHelper.swift
import Locanara
import NitroModules

/// Decodes Nitro option structs into Locanara chain constructor parameters
@available(iOS 15.0, macOS 14.0, *)
enum OndeviceAiHelper {

    // MARK: - Summarize

    static func bulletCount(from options: NitroSummarizeOptions?) -> Int {
        guard let outputType = options?.outputType else { return 1 }
        switch outputType {
        case .twoBullets: return 2
        case .threeBullets: return 3
        default: return 1
        }
    }

    static func inputType(from options: NitroSummarizeOptions?) -> String {
        guard let inputType = options?.inputType else { return "text" }
        switch inputType {
        case .conversation: return "conversation"
        default: return "text"
        }
    }

    // MARK: - Classify

    static func classifyOptions(from options: NitroClassifyOptions?) -> (categories: [String], maxResults: Int) {
        let categories: [String]
        if case .second(let arr) = options?.categories {
            categories = arr
        } else {
            categories = ["positive", "negative", "neutral"]
        }
        let maxResults: Int
        if case .second(let v) = options?.maxResults {
            maxResults = Int(v)
        } else {
            maxResults = 3
        }
        return (categories, maxResults)
    }

    // MARK: - Extract

    static func entityTypes(from options: NitroExtractOptions?) -> [String] {
        if case .second(let arr) = options?.entityTypes {
            return arr
        }
        return ["person", "location", "date", "organization"]
    }

    // MARK: - Chat

    static func chatOptions(from options: NitroChatOptions?) -> (systemPrompt: String, memory: (any Memory)?) {
        let systemPrompt: String
        if case .second(let s) = options?.systemPrompt {
            systemPrompt = s
        } else {
            systemPrompt = "You are a friendly, helpful assistant."
        }

        var memory: (any Memory)? = nil
        if case .second(let history) = options?.history, !history.isEmpty {
            memory = PrefilledMemory(history: history)
        }

        return (systemPrompt, memory)
    }

    // MARK: - Translate

    static func translateOptions(from options: NitroTranslateOptions) -> (sourceLanguage: String, targetLanguage: String) {
        (options.sourceLanguage, options.targetLanguage)
    }

    // MARK: - Rewrite

    static func rewriteStyle(from options: NitroRewriteOptions) -> RewriteOutputType {
        switch options.outputType {
        case .elaborate: return .elaborate
        case .emojify: return .emojify
        case .shorten: return .shorten
        case .friendly: return .friendly
        case .professional: return .professional
        case .rephrase: return .rephrase
        }
    }
}

// MARK: - Prefilled Memory

/// Memory adapter that provides pre-filled chat history from JS.
@available(iOS 15.0, macOS 14.0, *)
final class PrefilledMemory: Memory, @unchecked Sendable {
    private let entries: [MemoryEntry]

    init(history: [NitroChatMessage]) {
        self.entries = history.map { msg in
            MemoryEntry(role: msg.role.stringValue, content: msg.content)
        }
    }

    func load(for input: ChainInput) async -> [MemoryEntry] { entries }
    func save(input: ChainInput, output: ChainOutput) async { }
    func clear() async { }

    var estimatedTokenCount: Int {
        entries.reduce(0) { $0 + ($1.content.count / 4) }
    }
}
